pkg/urbit/noun/events.c

```diff
@@ -1,88 +1,101 @@
+///   - patch control (u3e_control control.bin): patch metadata, watermarks,
+///     and indices/mugs for pages in patch memory.
+///
+/// ### initialization (u3e_live())
+///
+///   - with the loom already mapped, all pages are marked dirty in a bitmap.
+///   - if snapshot is missing or partial, empty segments are created.
+///   - if a patch is present, it's applied (crash recovery).
+///   - snapshot segments are copied onto the loom; all included pages
+///     are marked clean and protected (read-only).
+///
+/// #### page faults (u3e_fault())
+///
+///   - stores into protected pages generate faults (currently SIGSEGV,
+///     handled outside this module).
+///   - faults are handled by dirtying the page and switching protections to
+///     read/write.
+///   - a guard page is initially placed in the approximate middle of the free
+///     space between the heap and stack at the time of the first page fault.
+///     when a fault is detected in the guard page, the guard page is recentered
+///     in the free space of the current road. if the guard page cannot be
+///     recentered, then memory exhaustion has occurred.
+///
+/// ### updates (u3e_save())
+///
+///   - all updates to a snapshot are made through a patch.
+///   - high/low watermarks for the north/south segments are established,
+///     and dirty pages below/above them are added to the patch.
+///     - modifications have been caught by the fault handler.
+///     - newly-used pages are automatically included (preemptively dirtied).
+///     - unused, innermost pages are reclaimed (segments are truncated to the
+///       high/low watermarks; the last page in each is always adjacent to the
+///       contiguous free space).
+///   - patch pages are written to memory.bin, metadata to control.bin.
+///   - the patch is applied to the snapshot segments, in-place.
+///   - patch files are deleted.
+///
+/// ### limitations
+///
+///   - loom page size is fixed (16 KB), and must be a multiple of the
+///     system page size. (can the size vary at runtime give south.bin's
+///     reversed order? alternately, if system page size > ours, the fault
+///     handler could dirty N pages at a time.)
+///   - update atomicity is suspect: patch application must either
+///     completely succeed or leave on-disk segments intact. unapplied
+///     patches can be discarded (triggering event replay), but once
+///     patch application begins it must succeed (can fail if disk is full).
+///     may require integration into the overall signal-handling regime.
+///   - any errors are handled with assertions; failed/partial writes are not
+///     retried.
+///
+/// ### enhancements
+///
+///   - use platform specific page fault mechanism (mach rpc, userfaultfd, &c).
+///   - implement demand paging / heuristic page-out.
+///   - add a guard page in the middle of the loom to reactively handle stack
+///     overflow.
+///   - parallelism
+///
```

Comment:
The only change here was from use of `//!` to `///`. `///` behaves better with auto-indent in `neovim` (and presumably other editors).